### PR TITLE
Add support for default implementation of static virtuals with method constraints

### LINKED ISSUE #73658: VerificationException with reflection, static abstract interface methods, and generic constraints
This:
```C#
internal static class Program
{
    static void Main()
    {
        Type s = typeof(Program).Assembly.GetType("Outer`1+Inner");
        Type i = s.GetInterface("IStatics");
        s.GetInterfaceMap(i);
    }
}

class Outer<TOuter>
{
    struct Inner : IStatics
    {
        public static void M<TInner>() where TInner : IConstraint { }
    }

    interface IStatics
    {
        static abstract void M<TInner>() where TInner : IConstraint;
    }

    interface IConstraint { }
}
```
throws:
```
Unhandled exception. System.Security.VerificationException: Method Outer`1+Inner[TOuter].M: type argument 'TInner' violates the constraint of type parameter 'TInner'.
   at System.RuntimeType.GetInterfaceMap(Type ifaceType)
```
Is that expected?

### LINKED ISSUE #78865: VerificationException static virtual interface methods, and generic constraints
### Description

I found code that throws a VerificationException for static virtual interface methods, and generic constraints, similar to this issue https://github.com/dotnet/runtime/issues/73658 but without reflection.

### Reproduction Steps

Run this code:

```
public interface ITestItem<T>
{
}

public interface IStaticInterfaceBase<T, in TRequest>
    where T : class
    where TRequest : IStaticInterfaceBase<T, TRequest>
{
    static abstract void TryInvoke<TItem>(TItem item, TRequest request) where TItem : ITestItem<T>;
}

public interface IStaticInterface<T, in TRequest> : IStaticInterfaceBase<T, TRequest>
    where T : class
    where TRequest : IStaticInterface<T, TRequest>
{
    static void IStaticInterfaceBase<T, TRequest>.TryInvoke<TItem>(TItem item, TRequest request)
    {
    }
}

public class Request : IStaticInterface<object, Request>
{
}

internal class Program
{
    public static void Invoke<T, TRequest>(TRequest request)
        where T : class
        where TRequest : IStaticInterfaceBase<T, TRequest> =>
        TRequest.TryInvoke((ITestItem<T>) null!, request);

    public static void Main(string[] args) => Invoke<object, Request>(new Request());
}
```

### Expected behavior

The code will run without error.

### Actual behavior

Unhandled exception. System.Security.VerificationException: Method StaticInterfaceBug.IStaticInterface`2[System.Object,StaticInterfaceBug.Request].StaticInterfaceBug.IStaticInterfaceBase<T,TRequest>.TryInvoke: type argument 'TItem' violates the constraint of type parameter 'TItem'.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by davidwrighton**
- The major problem was the logic which incorrectly would instantiate the methods when it wasn't necessary
- As the number of flags to the implementation functions has grown very large, this change also includes logic converting them all to a single flags variable when passing them around

Fixes #73658
Fixes #78865

### CODE REVIEW

File: src/coreclr/vm/methodtable.h [hunk 1]
```diff
@@ -63,6 +63,73 @@ class ClassFactoryBase;
...
+enum class ResolveVirtualStaticMethodFlags
+{
+    None = 0,
+    AllowNullResult = 1,
+    VerifyImplemented = 2,
+    AllowVariantMatches = 4,
+    InstantiateResultOverFinalMethodDesc = 8
+};
+
+inline ResolveVirtualStaticMethodFlags operator& (ResolveVirtualStaticMethodFlags left, ResolveVirtualStaticMethodFlags right)
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/methodtable.h:75]: Can we create a template for these? This is such a common pattern and we should just have it. Win32 has a simple non-template mechanism we can use as a starting point. Making this an actual template would be incredibly helpful.

```c++
#define DEFINE_ENUM_FLAG_OPERATORS(ENUMTYPE) \
extern "C++" { \
inline _ENUM_FLAG_CONSTEXPR ENUMTYPE operator | (ENUMTYPE a, ENUMTYPE b) WIN_NOEXCEPT { return ENUMTYPE(((_ENUM_FLAG_SIZED_INTEGER<ENUMTYPE>::type)a) | ((_ENUM_FLAG_SIZED_INTEGER<ENUMTYPE>::type)b)); } \
inline ENUMTYPE &operator |= (ENUMTYPE &a, ENUMTYPE b) WIN_NOEXCEPT { return (ENUMTYPE &)(((_ENUM_FLAG_SIZED_INTEGER<ENUMTYPE>::type &)a) |= ((_ENUM_FLAG_SIZED_INTEGER<ENUMTYPE>::type)b)); } \
inline _ENUM_FLAG_CONSTEXPR ENUMTYPE operator & (ENUMTYPE a, ENUMTYPE b) WIN_NOEXCEPT { return ENUMTYPE(((_ENUM_FLAG_SIZED_INTEGER<ENUMTYPE>::type)a) & ((_ENUM_FLAG_SIZED_INTEGER<ENUMTYPE>::type)b)); } \
inline ENUMTYPE &operator &= (ENUMTYPE &a, ENUMTYPE b) WIN_NOEXCEPT { return (ENUMTYPE &)(((_ENUM_FLAG_SIZED_INTEGER<ENUMTYPE>::type &)a) &= ((_ENUM_FLAG_SIZED_INTEGER<ENUMTYPE>::type)b)); } \
inline _ENUM_FLAG_CONSTEXPR ENUMTYPE operator ~ (ENUMTYPE a) WIN_NOEXCEPT { return ENUMTYPE(~((_ENUM_FLAG_SIZED_INTEGER<ENUMTYPE>::type)a)); } \
inline _ENUM_FLAG_CONSTEXPR ENUMTYPE operator ^ (ENUMTYPE a, ENUMTYPE b) WIN_NOEXCEPT { return ENUMTYPE(((_ENUM_FLAG_SIZED_INTEGER<ENUMTYPE>::type)a) ^ ((_ENUM_FLAG_SIZED_INTEGER<ENUMTYPE>::type)b)); } \
inline ENUMTYPE &operator ^= (ENUMTYPE &a, ENUMTYPE b) WIN_NOEXCEPT { return (ENUMTYPE &)(((_ENUM_FLAG_SIZED_INTEGER<ENUMTYPE>::type &)a) ^= ((_ENUM_FLAG_SIZED_INTEGER<ENUMTYPE>::type)b)); } \
}
```

File: src/coreclr/inc/enum_class_flags.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+template <typename T>
+inline constexpr auto operator& (T left, T right) -> decltype(T::support_use_as_flags)
+{
+    return static_cast<T>(static_cast<int>(left) & static_cast<int>(right));
+}
+
+template <typename T>
+inline constexpr auto operator| (T left, T right) -> decltype(T::support_use_as_flags)
+{
+    return static_cast<T>(static_cast<int>(left) & static_cast<int>(right));
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/inc/enum_class_flags.h:16]:
```suggestion
    return static_cast<T>(static_cast<int>(left) | static_cast<int>(right));
```

File: src/coreclr/inc/enum_class_flags.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+template <typename T>
+inline constexpr auto operator^ (T left, T right) -> decltype(T::support_use_as_flags)
+{
+    return static_cast<T>(static_cast<int>(left) & static_cast<int>(right));
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/inc/enum_class_flags.h:22]:
```suggestion
    return static_cast<T>(static_cast<int>(left) ^ static_cast<int>(right));
```

File: src/coreclr/inc/enum_class_flags.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+inline constexpr auto operator |= (T& left, T right) -> const decltype(T::support_use_as_flags)&
+{
+    left = left | right;
+    return left;
+}
+
+template <typename T>
+inline constexpr auto operator &= (T& left, T right) -> const decltype(T::support_use_as_flags)&
+{
+    left = left | right;
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/inc/enum_class_flags.h:41]:
```suggestion
    left = left & right;
```

File: src/coreclr/inc/enum_class_flags.h [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    return left;
+}
+
+template <typename T>
+inline constexpr auto operator ^= (T& left, T right) -> const decltype(T::support_use_as_flags)&
+{
+    left = left | right;
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/inc/enum_class_flags.h:48]:
```suggestion
    left = left ^ right;
```

File: src/tests/issues.targets [hunk 6]
```diff
@@ -694,6 +694,9 @@
 
     <!-- NativeAOT specific -->
     <ItemGroup Condition="'$(XunitTestBinBase)' != '' and '$(TestBuildMode)' == 'nativeaot' and '$(RuntimeFlavor)' == 'coreclr'">
+        <ExcludeList Include="$(XunitTestBinBase)/Loader/classloader/StaticVirtualMethods/RegressionTests/GetInterfaceMapWithStaticVirtualsAndConstraints/**">
+            <Issue>https://github.com/dotnet/runtime/issues/89122</Issue>
```
Comment (MichalStrehovsky) [inline src/tests/issues.targets:698]: We can track this in the "`GetInterfaceMap` doesn't really work" uber issue. There's already a bunch of tests blocked on it that we can ignore forever (or fix if there's ever a good enough business justification).

```suggestion
            <Issue>https://github.com/dotnet/runtime/issues/89157</Issue>
```

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by Joe4evr** [comment]
Did this not get into preview 7 somehow? I've ran the code that previously encountered #73658 and I'm still getting a `VerificationException` when calling `GetInterfaceMap`.

Additionally, in trying to make a minimized repo I'm encountering more bugs around getting a target method through reflection. Should I open a new issue for these? @AaronRobinsonMSFT @davidwrighton

**Comment by AaronRobinsonMSFT** [comment]
@Joe4evr Can you try the nightly from https://github.com/dotnet/installer#table. I'm not sure if these fixes made it into preview 7.

**Comment by AaronRobinsonMSFT** [comment]
If you are still hitting issues after trying the nightly, then please open a new issue.

**Comment by Joe4evr** [comment]
> @Joe4evr Can you try the nightly from https://github.com/dotnet/installer#table. I'm not sure if these fixes made it into preview 7.

Ah, thanks for that. Looks like I was concerned for nothing. That rc1 build *does* work in all the scenarios I tested.

**Pull request merged**